### PR TITLE
Kh/fix coverage on next project

### PR DESCRIPTION
**Pull request by KarimHmaissi**
Adding coverage in terminal and auto generated coverage report on running NX test command. Modifies the next generator to update the project.json config

### CODE REVIEW

File: packages/next/src/generators/init/utils/createCustomTestConfig.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                './**/*.{js,jsx,ts,tsx}',
+                './!**/.next/**',
+                './!**/*.d.ts',
+                './!**/*.config.*',
+                './!**/_app.*',
+            ],
+            codeCoverage: true,
+        },
+        configurations: {
+            ci: {
```
Comment (cloudratha) [inline packages/next/src/generators/init/utils/createCustomTestConfig.ts:22]: I think we should only run codeCoverage in ci by default.

File: packages/next/src/generators/init/utils/createCustomTestConfig.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                ci: true,
+            },
+        },
+    };
+}
+
+export default async function addCustomTestConfig(
```
Comment (cloudratha) [inline packages/next/src/generators/init/utils/createCustomTestConfig.ts:29]: Naming consistency to filename

File: packages/next/src/generators/init/utils/createCustomTestConfig.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import { joinPathFragments, Tree, updateJson } from '@nrwl/devkit';
```
Comment (cloudratha) [inline packages/next/src/generators/init/utils/createCustomTestConfig.ts:1]: File naming should be kebab-case for consistency.

File: packages/next/src/generators/init/utils/createCustomTestConfig.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    tree: Tree,
+    projectSourceRoot: string,
+) {
+    updateJson(
+        tree,
+        joinPathFragments(projectSourceRoot, 'project.json'),
+        json => {
+            const updatedJson = json;
+            updatedJson.targets = updatedJson.targets ?? {};
+            updatedJson.targets.test = getTestConfig(projectSourceRoot);
```
Comment (cloudratha) [inline packages/next/src/generators/init/utils/createCustomTestConfig.ts:39]: Should we merge the configs with the existing test executor options?

Comment (KarimHmaissi) [reply packages/next/src/generators/init/utils/createCustomTestConfig.ts:39]: Thats a good shout. Guess only really need to merge the CI coverage config with whats there already. Will update

File: packages/next/src/generators/init/utils/createCustomTestConfig.ts (hunk 3 above)
Comment (KarimHmaissi) [reply packages/next/src/generators/init/utils/createCustomTestConfig.ts:1]: This is my bad. Will update

File: packages/next/src/generators/init/utils/createCustomTestConfig.ts (hunk 2 above)
Comment (KarimHmaissi) [reply packages/next/src/generators/init/utils/createCustomTestConfig.ts:29]: This is my bad. Will update

**Comment by KarimHmaissi** [comment]
Added a utility to the core library to merge project.json target configs. Also updated requested file names and ensured coverage is reported only in CI. While this change will only effect the next plugin I believe it should go in as it will add the utilities and further work can be done to ensure all plugins have the required testing config.

File: packages/common/core/src/lib/project-config.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        (accumulator, source) =>
+            Object.keys(source).reduce(coalesceByKey(source), accumulator),
+        target,
+    );
+}
+
+function merge(a: unknown, b: unknown): unknown {
+    return isObject(a) && isObject(b)
+        ? deepMerge(a as Record<string, unknown>, b as Record<string, unknown>)
+        : isObject(a) && !isObject(b)
```
Comment (cloudratha) [inline packages/common/core/src/lib/project-config.ts:34]: This doesn't seem to handle deepMerging arrays? But rather prefers which ever is an object?

**Comment by KarimHmaissi** [comment]
Updated the merge with a combination of an NPM package [https://github.com/TehShrike/deepmerge ]( https://github.com/TehShrike/deepmerge ) and a custom array merge algorithm. It will now deep merge complex objects with nested arrays. Added the ability to specify a key in an object so the algorithm will treat that object differently. For example {files: '.tsx'}, {files: '.tsx'} will be merged together but {files: '.jsx'} will not. This will handle the cases such as merging the ESLint config overrides

**Review by cloudratha — APPROVED** [review]

**Pull request merged**